### PR TITLE
#363 - Introduce new `/uma-rs-modify` command to be able modify existing resource 

### PR DESCRIPTION
**Pull request by duttarnab**
#363 - Introduce new `/uma-rs-modify` command to be able modify existing resource 
https://github.com/GluuFederation/oxd/issues/363

### CODE REVIEW

File: oxd-common/src/main/java/org/gluu/oxd/common/ErrorResponseCode.java [hunk 1]
```diff
@@ -82,7 +82,8 @@
     FAILED_TO_REMOVE_SITE(500, "remove_site_failed", "Failed to remove site."),
     REDIRECT_URI_IS_NOT_REGISTERED(400, "redirect_uri_is_not_registered", "The authorization redirect uri is not registered."),
     FAILED_TO_GET_DISCOVERY(500, "failed_to_get_discovery", "Failed to get OP discovery configuration."),
-    SSL_HANDSHAKE_ERROR(500, "ssl_handshake_error", "Unable to find valid certification path to requested target. Please check if key_store_path in oxd configuration is correct.");
+    SSL_HANDSHAKE_ERROR(500, "ssl_handshake_error", "Unable to find valid certification path to requested target. Please check if key_store_path in oxd configuration is correct."),
+    NO_UMA_OXD_ID_PARAMETER(400, "invalid_oxd_id_parameter", "oxd_id parameter is not specified or otherwise is not valid"),;
```
Comment (yuriyz) [inline oxd-common/src/main/java/org/gluu/oxd/common/ErrorResponseCode.java]: We already have `BAD_REQUEST_NO_OXD_ID`, no need to duplicate it.

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        update(umaResource, site);
+
+        return new RsModifyResponse(site.getOxdId());
+    }
+
+    private void setResource(org.gluu.oxd.server.model.UmaResource umaResource, UmaResource opUmaResource) {
+        umaResource.setScopes(opUmaResource.getScopes());
+        if (!Strings.isNullOrEmpty(opUmaResource.getScopeExpression()) && !opUmaResource.getScopeExpression().equals("null")) {
+            umaResource.setScopeExpressions(Lists.newArrayList());
+            umaResource.getScopeExpressions().add(opUmaResource.getScopeExpression());
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: umaResource.setScopeExpressions(Lists.newArrayList(opUmaResource.getScopeExpression()));

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        UmaResource opUmaResource = getResource(resourceService, params, umaResource.getId());
+        setResource(umaResource, opUmaResource);
+
+        try {
+            String pat = getUmaTokenService().getPat(params.getOxdId()).getToken();
+            resourceService.updateResource("Bearer " + pat, umaResource.getId(), opUmaResource);
+        } catch (ClientResponseFailure e) {
+            LOG.debug("Failed to update resource. Entity: " + e.getResponse().getEntity(String.class) + ", status: " + e.getResponse().getStatus(), e);
+            if (e.getResponse().getStatus() == 400 || e.getResponse().getStatus() == 401) {
+                LOG.debug("Try maybe PAT is lost on AS, force refresh PAT and re-try ...");
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: It seems something is missed here. Log says that oxd will try to refresh PAT but where is actual refreshing?

```
getUmaTokenService().obtainPat(params.getOxdId());
```

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        UmaMetadata discovery = getDiscoveryService().getUmaDiscoveryByOxdId(params.getOxdId());
+        UmaResourceService resourceService = UmaClientFactory.instance().createResourceService(discovery, getHttpService().getClientExecutor());
+
+        UmaResource opUmaResource = getResource(resourceService, params, umaResource.getId());
+        setResource(umaResource, opUmaResource);
+
+        try {
+            String pat = getUmaTokenService().getPat(params.getOxdId()).getToken();
+            resourceService.updateResource("Bearer " + pat, umaResource.getId(), opUmaResource);
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: In this line we update resource on OP with `opUmaResource` but on line 91 with `umaResource`. We don't care about `umaResource`, stick to single object (`opUmaResource`). `setResource` looks redundant at all.

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        umaResource.setScopes(params.getScopes());
+        umaResource.setType(umaResourceWithId.getType());
+        if (!Strings.isNullOrEmpty(params.getScopeExpression()) && !params.getScopeExpression().equals("null")) {
+            umaResource.setScopeExpression(params.getScopeExpression());
+        }
+
+        return umaResource;
+    }
+
+    private void update(org.gluu.oxd.server.model.UmaResource resource, Rp site) throws IOException {
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: updateRp

File: oxd-common/src/main/java/org/gluu/oxd/common/ErrorResponseCode.java (hunk 1 above)
Comment (duttarnab) [reply oxd-common/src/main/java/org/gluu/oxd/common/ErrorResponseCode.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java (hunk 2 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java (hunk 3 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java (hunk 4 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java (hunk 5 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        UmaResource opUmaResource = getResource(resourceService, params, umaResource.getId());
+
+        try {
+            String pat = getUmaTokenService().getPat(params.getOxdId()).getToken();
+            resourceService.updateResource("Bearer " + pat, umaResource.getId(), opUmaResource);
+        } catch (ClientResponseFailure e) {
+            LOG.debug("Failed to update resource. Entity: " + e.getResponse().getEntity(String.class) + ", status: " + e.getResponse().getStatus(), e);
+            if (e.getResponse().getStatus() == 400 || e.getResponse().getStatus() == 401) {
+                LOG.debug("Try maybe PAT is lost on AS, force refresh PAT and re-try ...");
+                resourceService.updateResource("Bearer " + getUmaTokenService().obtainPat(params.getOxdId()), umaResource.getId(), opUmaResource); //refresh pat and update
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Here we refreshed PAT and updateResource(). Lets assume that it works successfully. And then on next line we will `throw e` (line 89). Looks wrong, no? We should update RP and return response if refresh of PAT and further update is successful.  Create method where when you update resource you also update RP and return response. In this case you will never create bug that resource is updated on OP but not on db or error is thrown even if update operation was successful.

```
public RsModifyResponse update(String pat, Rp rp, UmaResourceService resourceService, UmaResource resource) {
    resourceService.updateResource("Bearer " + pat, resource.getId(), resource);
    updateRp(opUmaResource, rp, umaResource.getId());

    return new RsModifyResponse(rp.getOxdId());
}
```

Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/op/RsModifyOperation.java]: Yeah, it was a mistake. Now, corrected the error.

**Pull request merged**